### PR TITLE
changed: split HDF5 class in 3 parts

### PR DESCRIPTION
**Pull request by akva2**
- HDF5Writer writes results
- HDF5Reader reads
- HDF5Restart reads and writes restart data

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=106 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=106 ifem-navierstokes=93 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=106 ifem-navierstokes=93 ifem-boussinesq=6 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=106 ifem-navierstokes=93 ifem-boussinesq=6 ifem-elasticity=65 ifem-beamex=16 ifem-finitedeformation=16 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=106 ifem-navierstokes=93 ifem-boussinesq=6 ifem-elasticity=65 ifem-beamex=16 ifem-finitedeformation=16 ifem-fsi=33 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=106 ifem-navierstokes=93 ifem-boussinesq=6 ifem-elasticity=65 ifem-beamex=16 ifem-finitedeformation=16 ifem-fsi=33 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=106 ifem-navierstokes=93 ifem-boussinesq=6 ifem-elasticity=65 ifem-beamex=16 ifem-finitedeformation=16 ifem-fsi=33 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=106 ifem-navierstokes=93 ifem-boussinesq=6 ifem-elasticity=65 ifem-beamex=16 ifem-finitedeformation=16 ifem-fsi=33 please

File: Apps/Common/SIMSolver.h [hunk 1]
```diff
@@ -45,16 +46,15 @@ template<class T1> class SIMSolverStat : public SIMadmin
   //! \brief Handles application data output.
   //! \param[in] hdf5file The file to save to
   //! \param[in] saveInterval The stride in the output file
-  //! \param[in] restartInterval The stride in the restart file
   void handleDataOutput(const std::string& hdf5file,
-                        int saveInterval = 1, int restartInterval = 0)
+                        int saveInterval = 1)
```
Comment (kmokstad) [inline Apps/Common/SIMSolver.h]: move `saveInterval = 1)` to line 49

File: Apps/Common/SIMSolver.h [hunk 2]
```diff
@@ -123,10 +123,14 @@ template<class T1> class SIMSolver : public SIMSolverStat<T1>
...
+    restartAdm = nullptr;
   }
 
-  //! \brief Empty destructor.
-  virtual ~SIMSolver() {}
+  //! \brief The destructor closes the restart data handler.
+  virtual ~SIMSolver()
+  {
+    delete restartAdm;
+  }
```
Comment (kmokstad) [inline Apps/Common/SIMSolver.h]: the destructor can still be on a single line. And in the doxy, maybe `deletes` is more correct that `closes` ?

File: Apps/Common/SIMSolver.h [hunk 3]
```diff
@@ -204,12 +221,12 @@ template<class T1> class SIMSolver : public SIMSolverStat<T1>
...
-          this->serialize(data))
-        return SIMSolverStat<T1>::exporter->dumpTimeLevel(&tp,newMesh,&data);
-      else // no restart dump, or serialization failure
-        return SIMSolverStat<T1>::exporter->dumpTimeLevel(&tp,newMesh);
+      HDF5Restart::SerializeData data;
+      bool result = true;
+      if (restartAdm && restartAdm->dumpStep(tp) && this->serialize(data))
+        result = restartAdm->writeData(tp,data);
+
+      return result && SIMSolverStat<T1>::exporter->dumpTimeLevel(&tp,newMesh);
```
Comment (kmokstad) [inline Apps/Common/SIMSolver.h]: Don't need the `result`temporary if you write it like this (same logic):

```
      HDF5Restart::SerializeData data;
      if (restartAdm && restartAdm->dumpStep(tp) && this->serialize(data))
        if (!restartAdm->writeData(tp,data))
          return false;

      return SIMSolverStat<T1>::exporter->dumpTimeLevel(&tp,newMesh);
```

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=106 ifem-navierstokes=93 ifem-boussinesq=6 ifem-elasticity=65 ifem-beamex=16 ifem-finitedeformation=16 ifem-fsi=33 please

File: Apps/Common/SIMSolver.h [hunk 4]
```diff
@@ -161,16 +161,29 @@ template<class T1> class SIMSolver : public SIMSolverStat<T1>
     return 0;
   }
 
+  //! \brief Handles application data output.
+  //! \param[in] hdf5file The file to save to
+  //! \param[in] saveInterval The stride in the output file
+  //! \param[in] restartInterval The stride in the restart file
+  void handleDataOutput(const std::string& hdf5file,
+                           int saveInterval = 1, int restartInterval = 0)
```
Comment (kmokstad) [inline Apps/Common/SIMSolver.h]: fix indent

**Pull request merged**